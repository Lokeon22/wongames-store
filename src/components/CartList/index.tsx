import { useCart } from "../../hooks/use-cart"
import Link from "next/link"
import Button from "../Button"
import GameItem from "../GameItem"
import Empty from "../Empty"
import Loader from "../Loader"

import * as S from "./styles"

export type CartListProps = {
  hasButton?: boolean
}

const CartList = ({ hasButton = false }: CartListProps) => {
  const { items, total, loading } = useCart()

  if (loading)
    return (
      <S.Loading>
        <Loader />
      </S.Loading>
    )

  return (
    <S.Wrapper isEmpty={!items}>
      {items?.length ? (
        <>
          <S.GameList>
            {items &&
              items.map((item) => <GameItem key={item.title} {...item} />)}
          </S.GameList>
          <S.Footer>
            {!hasButton && <span>Total:</span>}
            <S.Total>{total}</S.Total>

            {hasButton && (
              <Link href="/cart">
                <Button>Buy it now</Button>
              </Link>
            )}
          </S.Footer>
        </>
      ) : (
        <Empty
          title="Your cart is empty"
          description="Go back to the store and explore great games"
        />
      )}
    </S.Wrapper>
  )
}

export default CartList
