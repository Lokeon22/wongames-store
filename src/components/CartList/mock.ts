const items = [
  {
    id: "1",
    img: "https://source.unsplash.com/user/willianjusten/151x70",
    title: "Red Dead Redemption 2",
    slug: "red-dead",
    price: "R$ 215,00"
  },
  {
    id: "2",
    img: "https://source.unsplash.com/user/willianjusten/151x70",
    title: "Borderlands 3",
    slug: "borderlands-3",
    price: "R$ 215,00"
  }
]

export default items
