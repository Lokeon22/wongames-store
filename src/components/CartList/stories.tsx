import { StoryFn, Meta } from "@storybook/react"
import CartList from "."

import items from "./mock"

export default {
  title: "CartList",
  component: CartList,
  argTypes: {
    carContextValue: {
      type: "string"
    }
  },
  parameters: {
    backgrounds: {
      default: "won-dark"
    }
  }
} as Meta

export const Default: StoryFn = (args) => (
  <div style={{ maxWidth: "80rem" }}>
    <CartList {...args} />
  </div>
)

Default.args = {
  total: "R$ 330,00",
  cartContextValue: { items }
}

export const WithButton: StoryFn = (args) => (
  <div style={{ maxWidth: "80rem" }}>
    <CartList {...args} hasButton />
  </div>
)

WithButton.args = {
  total: "R$ 330,00",
  cartContextValue: { items }
}

export const Empty: StoryFn = () => (
  <div style={{ maxWidth: "80rem" }}>
    <CartList />
  </div>
)
