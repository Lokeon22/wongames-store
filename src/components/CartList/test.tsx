import { screen } from "@testing-library/react"
import { renderWithTheme } from "../../utils/tests/helpers"

import CartList from "."
import cartMock from "./mock"

describe("<CartList />", () => {
  it("should render the cartList", () => {
    const { container } = renderWithTheme(
      <CartList items={cartMock} total="R$ 330,00" />
    )

    expect(screen.getAllByRole("heading")).toHaveLength(2)
    expect(screen.getByText("R$ 330,00")).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render the button", () => {
    renderWithTheme(<CartList items={cartMock} total="R$ 330,00" hasButton />)

    expect(screen.getByText(/buy it now/i)).toBeInTheDocument()
  })

  it("should render empty if there are no games", () => {
    renderWithTheme(<CartList />)

    expect(screen.getByText(/your cart is empty/i)).toBeInTheDocument()
  })
})
