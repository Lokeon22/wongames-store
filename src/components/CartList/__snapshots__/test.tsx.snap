// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> should render the cartList 1`] = `
.c2 {
  padding: 2.4rem;
  border-bottom: 0.1rem solid #EAEAEA;
  background-color: #FAFAFA;
}

.c3 {
  width: 100%;
  display: flex;
}

.c4 {
  flex-shrink: 0;
  margin-right: 1.2rem;
  width: 9.6rem;
  height: 5.6rem;
}

.c5 {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c6 {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: start;
}

.c7 {
  font-size: 1.4rem;
  line-height: 1.4rem;
  font-weight: 600;
  color: #030517;
  margin-bottom: 0.8rem;
}

.c9 {
  width: fit-content;
  color: #FAFAFA;
  padding: 0.2rem 0.8rem;
  font-size: 1.4rem;
  background-color: #3CD3C1;
  border-radius: 0.2rem;
}

.c8 {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c0 {
  background-color: #FAFAFA;
  display: flex;
  flex-direction: column;
  align-self: start;
}

.c10 {
  background: #eee;
  color: #030517;
  font-weight: 600;
  font-size: 1.4rem;
  padding: 2rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c11 {
  color: #F231A5;
}

.c1 {
  max-height: 40rem;
  overflow-y: auto;
}

@media (max-width: 1170px) {
  .c2 {
    padding: 0.8rem;
  }
}

@media (min-width: 768px) {
  .c4 {
    width: 15rem;
    height: 7rem;
    margin-right: 1.6rem;
  }
}

@media (min-width: 768px) {
  .c7 {
    font-size: 2.0rem;
    line-height: 2.0rem;
  }
}

@media (min-width: 768px) {
  .c10 {
    font-size: 1.6rem;
    padding: 2.4rem;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Red Dead Redemption 2"
            class="c5"
            src="https://source.unsplash.com/user/willianjusten/151x70"
          />
        </div>
        <div
          class="c6"
        >
          <h3
            class="c7"
          >
            Red Dead Redemption 2
          </h3>
          <div
            class="c8"
          >
            <p
              class="c9"
            >
              R$ 215,00
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Borderlands 3"
            class="c5"
            src="https://source.unsplash.com/user/willianjusten/151x70"
          />
        </div>
        <div
          class="c6"
        >
          <h3
            class="c7"
          >
            Borderlands 3
          </h3>
          <div
            class="c8"
          >
            <p
              class="c9"
            >
              R$ 215,00
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c10"
  >
    <span>
      Total:
    </span>
    <span
      class="c11"
    >
      R$ 330,00
    </span>
  </div>
</main>
`;
